Mathsteps demo

nitin42 opened the pull request:
Powered by React and bootstrapped with`create-react-app`!

evykassirer commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+      steps = mathsteps.solveEquation(value);
+    }
+
+    if (steps.length !== 0) {
```
just to be sure, `if (steps && steps.length !== 0)`

evykassirer commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+    this.setState({
+      simplify: e.target.checked,
+      solve: false,
+      value: "2x + 2x + x + x"
```
you have this string twice in the file - maybe define them as constants at the top?

```
const EXPRESSION_EXAMPLE = "2x + 2x + x + x" 
const EQUATION_EXAMPLE = "2x + 2x = 12"
```

nitin42 replied:
👍

evykassirer commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+    return (
+      <div>
+        <Container>
+          <Heading>Let's get started. Enter an equation {""}</Heading>
```
"Enter a math expression or equation."

what's `{""}` ?

nitin42 replied:
I formatted the code with `jsPrettier` and it added `{""}`. Anyway I've removed this !

evykassirer commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+      </div>
+    );
+  }
+}
```
I think this looks funky (highlighted weird) in the diff because the file extension isn't jsx (don't know for sure though)

evykassirer commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+import Steps from "./Steps";
+import { media } from "../utils/main";
+
+const Container = styled.div`
```
If we do styles in the same file as the components, can we put it at the bottom of the file instead of the top? It's nice to open a file and immediately see what it is

nitin42 replied:
I've removed styled-components as a dependency for the demo. Just css !!

evykassirer commented on src/components/Radio.js:
```diff
@@ -0,0 +1,28 @@
+          checked={simplify}
+          onChange={handleSimplify}
+        />
+        &nbsp;Simplify Expression
```
do you think you could use css instead of `&nbsp;`?

nitin42 replied:
Fixed!

evykassirer commented on src/App.css:
```diff
@@ -0,0 +1,13 @@
+  height: 100%;
+}
+
+.options {
```
why is some css in this file and some in the same file as the components?

nitin42 replied:
Done!

evykassirer commented on src/components/Search.js:
```diff
@@ -0,0 +1,29 @@
+let Search = ({ value, handleChange, handleClick }) => {
+	return (
+    <div className="field has-addons">
+      <p className="control">
```
is this classname used for anything?

nitin42 replied:
We rely on [Bulma](bulma.io) which is a CSS framework for adding some elements, improving typography and many more things. It lessen the burden of styling so that users can focus more on adding features.

evykassirer commented on src/components/Search.js:
```diff
@@ -0,0 +1,29 @@
+    <div className="field has-addons">
+      <p className="control">
+        <input
+          style={{ marginTop: "30px" }}
```
I'd rather you not put style directly in as an argument like this, use css styling in a separate file or in a separate part of the file like you've bee doing elsewhere

evykassirer commented on src/components/Show.js:
```diff
@@ -0,0 +1,34 @@
+  `}
+`;
+
+let Show = ({ renderer }) => {
```
I don't think this component is necessary, since most of what it's displaying you're just passing in - what do you think about just moving all this back to `Steps`?

nitin42 replied:
No need !! I follow single responsibility principle. Each component should only have one responsibility and only one reason to change. Moving this to `Steps` component may obfuscate the codebase later when we expand (substeps view and JSON we talked about).

evykassirer commented on src/components/Steps.js:
```diff
@@ -0,0 +1,53 @@
+        <div key={i} className="steps">
+          {simplify
+            ? <Simplify step={step} />
+            : solve ? <Solve step={step} i={i} /> : <div />}
```
I found this hard to read - how about:

```
{simplify && <Simplify step={step} />}
{solve && <Solve step={step} i={i} />}
```

nitin42 replied:
👍

evykassirer commented on src/components/Steps.js:
```diff
@@ -0,0 +1,53 @@
+      this.setState({
+        updateSteps: nextProps.steps
+      });
+      console.log(this.state.updateSteps);
```
no console.logs in prod code pls

nitin42 replied:
Sorry!

evykassirer commented on src/components/Steps.js:
```diff
@@ -0,0 +1,53 @@
+  };
+
+  state = {
+    updateSteps: []
```
I don't think you need this, you can just have `steps` in props and update the props passed into this component to have empty steps if the user switches modes

nitin42 replied:
Ohk! I've used `updateState` for avoiding a re-render of the whole component when the steps are same. Performance boost!

If you look carefully, I've also used `PureComponent` instead of 'Component'.

evykassirer commented on src/components/Simplify.js:
```diff
@@ -0,0 +1,23 @@
+  return this.charAt(0).toUpperCase() + this.slice(1);
+};
+
+let Simplify = ({ step }) => {
```
simplify also has multiple steps (maybe not in your example) and can so this file can probably be merged with the Solve component - there's only one or two small differences in how it works

Let me know if you need more info about simplify vs solve

nitin42 replied:
Yes! What more changes can I make ? I don't know the internals for `simplifying` method.

nitin42 commented on public/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "name": "Mathsteps working demo",
+  "icons": [
+    {
+      "src": "favicon.ico",
```
Sure! Just give me a link then.

nitin42 commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+> Note - The demo is bootstrapped using [`create-react-app`](https://github.com/facebookincubator/create-react-app).
+
+## Build
```
We have to bundle everything after we are done with the changes. Yes, before deploying !

nitin42 commented on README.md:
````diff
@@ -1,2 +1,50 @@
+$ npm run build
+```
+
+## Deploying 
````
Sure! We can switch to GitHub pages. Regarding the deploying process, I will add a note for that.

nitin42 commented on README.md:
```diff
@@ -1,2 +1,50 @@
+
+For domain name, use `mathsteps.surge.sh` only 😅
+
+Traversing the codebase should be quite easy. All the components live under [`components`](./src/components) directory.
```
Done!

nitin42 commented on package.json:
```diff
@@ -0,0 +1,26 @@
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
+    "mathsteps-test": "^1.0.2",
```
Oh! Something ahout `mathsteps-test` -
 
I added Babel and Webpack to my fork and published it locally on npm with a different name. But this will change as we land the PR which I created for adding Webpack to Mathsteps.

nitin42 commented on src/components/Show.js:
```diff
@@ -0,0 +1,34 @@
+`;
+
+let Show = ({ renderer }) => {
+	return (
```
Ok! My Sublime Text preferences changes every time with the new update.

nitin42 commented on src/components/Header.js:
```diff
@@ -0,0 +1,69 @@
+            GitHub
+          </a>
+        </NameSpace>
+        <Demo />
```
Yup! Done

nitin42 commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+import React, { Component } from "react";
```
What more functionality ?

nitin42 commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+  static displayName = "Demo";
+
+  state = {
+    value: "2x + 2x + x + x",
```
Ya that will be better!

nitin42 commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+  state = {
+    value: "2x + 2x + x + x",
+    steps: [],
+    simplify: true,
```
I'm not certain I follow though.. could you make that change once I push this PR?

nitin42 commented on src/components/Demo.js:
```diff
@@ -0,0 +1,104 @@
+    solve: false
+  };
+
+  handleChange = e => {
```
Good point!

nitin42 commented on src/components/Solve.js:
```diff
@@ -0,0 +1,25 @@
+      	Step {i + 1}
+    	</strong>
+    	<p>
+      	Previous expression : {step.oldEquation.print()}
```
👍

nitin42 commented:
Changes made : 
* Removed `style-components`. 
* Added comments.
* Code refactored.
* Individual CSS file for each component.

Changes need to be made :
* Add blueberry favicon.
* Change the deployment service (use GitHub page instead of surge).
* Update the README.

evykassirer commented:
just getting to this now - I don't think you pushed your changes! I don't see anything since my review

evykassirer commented:
or are you waiting until you make the rest of the changes? @aliang8 can you see if you can find the favicon?

nitin42 commented:
Waiting for favicon! 😅

evykassirer commented:
can you use this as the favicon?

https://github.com/socraticorg/atom/blob/cb5d772c0e270e0c827ac028c6037a512b1a3296/static/images/app/header-icons/new/blueberry-outline.png?raw=true

nitin42 commented:
Done !!

evykassirer commented on src/components/Simplify.js:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { trimAll, trimOne, normalize } from '../utils/main';
+
+String.prototype.capitalize = function() {
```
`String prototype is read only, properties should not be added  no-extend-native`

please just make a new function outside of String for this :)

rushilsrivastava replied:
`normalize` is imported, but never constructed and exported in [main.js](https://github.com/nitin42/mathsteps-demo/blob/master/src/utils/main.js#L27).

evykassirer commented on src/utils/main.js:
```diff
@@ -0,0 +1,17 @@
+const normalize = (step) => {
```
can you rename this to show more why we're using it? something like `convertChangeTypeToUserFriendlyString`

since it's only used in one particular place, you could also move it to that component

evykassirer commented on src/utils/main.js:
```diff
@@ -0,0 +1,17 @@
+	return step.toLowerCase().replace(/_/g, ' ');
+}
+
+const trimAll = (step) => {
```
why do you need these trim functions?

nitin42 replied:
To trim extra spaces and prettify the output! These are useful at some point

evykassirer commented on src/registerServiceWorker.js:
```diff
@@ -0,0 +1,51 @@
+// In production, we register a service worker to serve assets from local cache.
+
+// This lets the app load faster on subsequent visits in production, and gives
+// it offline capabilities. However, it also means that developers (and users)
```
I looked into a bit and found that some people have issues with it

https://github.com/facebookincubator/create-react-app/issues/2398

do we print really clearly when cached content is the content being loaded?

nitin42 replied:
Yup!

evykassirer commented on src/components/Show.js:
```diff
@@ -0,0 +1,15 @@
+import './Show.css';
+	
+// Show component for displaying the steps
+let Show = ({ renderer }) => {
```
please move the body of this into `Step.jsx`, it's unnecessary complication for the app right now

evykassirer commented on src/components/Radio.js:
```diff
@@ -0,0 +1,30 @@
+import './Radio.css';
+
+/** Radio component for switching between simplify and solve option*/
+let Radio = ({ simplify, handleSimplify, solve, handleSolve }) => {
```
can you put a `renderRadio` function in `Demo.jsx` that returns this content? I would rather only make components for things that are fairly complex on their own or are used more than once in the code. If something gets more complicated or gets used somewhere else, we can move it to its own file later

nitin42 commented:
Yes, you are right! It would be better if we wait for the changes in `mathsteps` because I am using a custom build for the demo!

evykassirer commented:
The rest of the changes aren't blocked on that though, so feel free to keep updating this PR before the webpack thing gets reviewed if you want :) then once webpack is in we can update this and deploy!

nitin42 commented:
Sounds good!

nitin42 commented on src/components/Simplify.js:
```diff
@@ -0,0 +1,23 @@
+	return (
+		<div>
+    	<p>
+      	<strong>Input expression:</strong>{' '}
```
For printing pretty output (default by `prettier`)

rushilsrivastava commented:
In [Main.js](https://github.com/nitin42/mathsteps-demo/blob/master/src/utils/main.js), you are missing the `normalize(str)` function. I assume it is something like:

```
const normalize = (str) => {
  return str.replace(/_/gi,' ').toLowerCase();
};

export {
	userFriendlyString,
	trimAll,
	trimOne,
	media,
	capitalize,
        normalize
};
```

evykassirer commented:
I'm in a rush for the demo so I've moved it here: https://github.com/evykassirer/mathsteps-website/pull/1

Feel free to make comments here! And we can move some of your ideas from this repo over there too after I merge the initial site (in the next two days)

evykassirer commented:
turns out I don't even have write access here anymore, so good thing I moved it to my personal. Consider this repo deprecated :/
